compose.py

```diff
@@ -0,0 +1,305 @@
+# (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+# OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+import argparse
+import os
+import subprocess
+import sys
+import docker
+
+### Global variables
+TRITON_VERSION_MAP = {
+    '2.11.0dev':
+      ('21.06dev',   # triton container
+       '21.05')      # upstream container     
+}
+
+EXAMPLE_BACKENDS = ['identity', 'square', 'repeat']
+CORE_BACKENDS = ['tensorrt', 'ensemble']
+NONCORE_BACKENDS = [
+    'tensorflow1', 'tensorflow2', 'onnxruntime', 'python', 'dali', 'pytorch',
+    'openvino', 'fil'
+]
+EXAMPLE_REPOAGENTS = ['checksum']
+FLAGS = None
+
+DEPENDENCY_MAP = {
+    'python': 
+    ('python3-pip'),
+    'pytorch':
+    ('python3-dev')
+}
+#### helper functions
+def log(msg, force=False):
+    if force or not FLAGS.quiet:
+        try:
+            print(msg, file=sys.stderr)
+        except Exception:
+            print('<failed to log>', file=sys.stderr)
+
+
+def log_verbose(msg):
+    if FLAGS.verbose:
+        log(msg, force=True)
+
+
+def fail(msg):
+    fail_if(True, msg)
+
+
+def fail_if(p, msg):
+    if p:
+        print('error: {}'.format(msg), file=sys.stderr)
+        sys.exit(1)
+
+
+##### create base image
+def make_dockerfile(ddir, dockerfile_name):
+    # Set enviroment variables
+    df = '''
+#
+# Multistage build.
+#
+FROM nvcr.io/nvidia/tritonserver:{}-py3 as full
+FROM nvcr.io/nvidia/tritonserver:{}-py3-min
+ARG TRITON_VERSION={}
+ARG TRITON_CONTAINER_VERSION={}
+ENV TRITON_SERVER_VERSION ${{TRITON_VERSION}}
+ENV NVIDIA_TRITON_SERVER_VERSION ${{TRITON_CONTAINER_VERSION}}
+ENV TRITON_SERVER_VERSION ${{TRITON_VERSION}}
+ENV NVIDIA_TRITON_SERVER_VERSION ${{TRITON_CONTAINER_VERSION}}
+LABEL com.nvidia.tritonserver.version="${{TRITON_SERVER_VERSION}}"
+ENV PATH /opt/tritonserver/bin:${{PATH}}
+'''.format(FLAGS.upstream_container_version, FLAGS.upstream_container_version, FLAGS.version, FLAGS.container_version)
+
+    # Copy over files
+    df += '''
+# Create a user that can be used to run triton as
+# non-root. Make sure that this user to given ID 1000. All server
+# artifacts copied below are assign to this user.
+ENV TRITON_SERVER_USER=triton-server
+RUN userdel tensorrt-server > /dev/null 2>&1 || true && \
+    if ! id -u $TRITON_SERVER_USER > /dev/null 2>&1 ; then \
+        useradd $TRITON_SERVER_USER; \
+    fi && \
+    [ `id -u $TRITON_SERVER_USER` -eq 1000 ] && \
+    [ `id -g $TRITON_SERVER_USER` -eq 1000 ]
+
+WORKDIR /opt/tritonserver
+RUN rm -fr /opt/tritonserver/*
+COPY --chown=1000:1000 --from=full /opt/tritonserver/LICENSE .
+COPY --chown=1000:1000 --from=full /opt/tritonserver/TRITON_VERSION .
+COPY --chown=1000:1000 --from=full /opt/tritonserver/NVIDIA_Deep_Learning_Container_License.pdf .
+COPY --chown=1000:1000 --from=full /opt/tritonserver/bin bin/
+COPY --chown=1000:1000 --from=full /opt/tritonserver/lib lib/
+COPY --chown=1000:1000 --from=full /opt/tritonserver/include/triton/core include/triton/core
```

Comment:
Wouldn't your `min` based image then still have everything from `full` image?

Reply:
I was wondering if I need to copy over the `bin/`, `lib/` or `include/triton/core`. I think  the benefit for this script is that it doesn't include some of the backends in the `full` image

Reply:
Min starts with nothing triton related. So the above copy over just the core parts of triton which are always required. The backends and repo-agents are the things that are included conditionally.